# Урок 2. Создание признакового пространства.

Для продолжения работы во 2 д з нужно выполнить д з к 1 уроку, что и сделано ниже. За ним идет уже работа ко 2 д з.

In [1]:
apostrophe_dict = {
"ain't": "am not / are not",
"aren't": "are not / am not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is",
"i'd": "I had / I would",
"i'd've": "I would have",
"i'll": "I shall / I will",
"i'll've": "I shall have / I will have",
"i'm": "I am",
"i've": "I have",
"isn't": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}



short_word_dict = {
"121": "one to one",
"a/s/l": "age, sex, location",
"adn": "any day now",
"afaik": "as far as I know",
"afk": "away from keyboard",
"aight": "alright",
"alol": "actually laughing out loud",
"b4": "before",
"b4n": "bye for now",
"bak": "back at the keyboard",
"bf": "boyfriend",
"bff": "best friends forever",
"bfn": "bye for now",
"bg": "big grin",
"bta": "but then again",
"btw": "by the way",
"cid": "crying in disgrace",
"cnp": "continued in my next post",
"cp": "chat post",
"cu": "see you",
"cul": "see you later",
"cul8r": "see you later",
"cya": "bye",
"cyo": "see you online",
"dbau": "doing business as usual",
"fud": "fear, uncertainty, and doubt",
"fwiw": "for what it's worth",
"fyi": "for your information",
"g": "grin",
"g2g": "got to go",
"ga": "go ahead",
"gal": "get a life",
"gf": "girlfriend",
"gfn": "gone for now",
"gmbo": "giggling my butt off",
"gmta": "great minds think alike",
"h8": "hate",
"hagn": "have a good night",
"hdop": "help delete online predators",
"hhis": "hanging head in shame",
"iac": "in any case",
"ianal": "I am not a lawyer",
"ic": "I see",
"idk": "I don't know",
"imao": "in my arrogant opinion",
"imnsho": "in my not so humble opinion",
"imo": "in my opinion",
"iow": "in other words",
"ipn": "I’m posting naked",
"irl": "in real life",
"jk": "just kidding",
"l8r": "later",
"ld": "later, dude",
"ldr": "long distance relationship",
"llta": "lots and lots of thunderous applause",
"lmao": "laugh my ass off",
"lmirl": "let's meet in real life",
"lol": "laugh out loud",
"ltr": "longterm relationship",
"lulab": "love you like a brother",
"lulas": "love you like a sister",
"luv": "love",
"m/f": "male or female",
"m8": "mate",
"milf": "mother I would like to fuck",
"oll": "online love",
"omg": "oh my god",
"otoh": "on the other hand",
"pir": "parent in room",
"ppl": "people",
"r": "are",
"rofl": "roll on the floor laughing",
"rpg": "role playing games",
"ru": "are you",
"shid": "slaps head in disgust",
"somy": "sick of me yet",
"sot": "short of time",
"thanx": "thanks",
"thx": "thanks",
"ttyl": "talk to you later",
"u": "you",
"ur": "you are",
"uw": "you’re welcome",
"wb": "welcome back",
"wfm": "works for me",
"wibni": "wouldn't it be nice if",
"wtf": "what the fuck",
"wtg": "way to go",
"wtgp": "want to go private",
"ym": "young man",
"gr8": "great"
}


emoticon_dict = {
":)": "happy",
":‑)": "happy",
":-]": "happy",
":-3": "happy",
":->": "happy",
"8-)": "happy",
":-}": "happy",
":o)": "happy",
":c)": "happy",
":^)": "happy",
"=]": "happy",
"=)": "happy",
"<3": "happy",
":-(": "sad",
":(": "sad",
":c": "sad",
":<": "sad",
":[": "sad",
">:[": "sad",
":{": "sad",
">:(": "sad",
":-c": "sad",
":-< ": "sad",
":-[": "sad",
":-||": "sad"
}

In [2]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import nltk
import warnings 
warnings.filterwarnings("ignore", category=DeprecationWarning)
import os

In [3]:
!pip install HTMLParser

In [4]:
def apostrophe_term_replacer(text,dictionary,blank_replacement=False):
    if len(text)==0:
        return text
    for word in text.split():
        if word in apostrophe_dict:
            replacement= dictionary[word]
            if '/' in replacement:
                replacement=replacement[0:replacement.index('/')].rstrip()
            if blank_replacement == True:
                text=text.replace(word,'')
            else:                
                text=text.replace(word,replacement)
    return text

In [5]:
def del_symbols_with_len(text,len_to_del=1):
    return ' '.join([w for w in text.split() if len(w)>len_to_del])

In [9]:
import nltk

nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Windows\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Windows\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\Windows\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [8]:
def get_wordnet_pos(word):
    """Map POS tag to first character lemmatize() accepts"""
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

In [10]:
from html.parser import HTMLParser
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.corpus import wordnet


stop_words = set(stopwords.words("english"))

train_df = pd.read_csv('train_tweets.csv')
test_df = pd.read_csv('test_tweets.csv')
combine_df = train_df.append(test_df, ignore_index = True, sort = False)
combine_df.reset_index(inplace=True)

html_parser = HTMLParser()
combine_df['clean_tweet']=''
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']= html_parser.unescape(combine_df.at[row, 'tweet'])

for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=re.sub('@[\w]*',' ', combine_df.at[row, 'clean_tweet'])
    
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=combine_df.at[row, 'clean_tweet'].lower()    

for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=apostrophe_term_replacer(combine_df.at[row, 'clean_tweet'],apostrophe_dict)

for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=apostrophe_term_replacer(combine_df.at[row, 'clean_tweet'],short_word_dict)

for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=apostrophe_term_replacer(combine_df.at[row, 'clean_tweet'],emoticon_dict,blank_replacement=True)
    
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=re.sub(r'[^\w\s]',' ', combine_df.at[row, 'clean_tweet'])
    
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=re.sub(r'[^a-zA-Z0-9]',' ', combine_df.at[row, 'clean_tweet'])
    
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=re.sub(r'[^a-zA-Z]',' ', combine_df.at[row, 'clean_tweet'])
    
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'clean_tweet']=del_symbols_with_len(combine_df.at[row, 'clean_tweet'])
    
combine_df['tweet_token']=''
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'tweet_token']=word_tokenize(combine_df.at[row, 'clean_tweet'])
    
combine_df['tweet_token_filtered']=''
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'tweet_token_filtered']=[word for word in combine_df.at[row, 'tweet_token'] if not word in stop_words]
    
stemmer = PorterStemmer()

combine_df['tweet_stemmed']=''
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'tweet_stemmed']=[stemmer.stem(word) for word in combine_df.at[row, 'tweet_token_filtered']]

lemmatizer = WordNetLemmatizer()

combine_df['tweet_lemmatized']=''
for row in np.arange(len(combine_df)):
    combine_df.at[row, 'tweet_lemmatized']=[lemmatizer.lemmatize(word, get_wordnet_pos(word)) for word in combine_df.at[row, 'tweet_token_filtered']]
    #combine_df.at[row, 'tweet_lemmatized']=[lemmatizer.lemmatize(word) for word in combine_df.at[row, 'tweet_token_filtered']]

import pickle
with open('combine_df.pickle', 'wb') as handle:
    pickle.dump(combine_df, handle, protocol=pickle.HIGHEST_PROTOCOL)
    
#проверка
combine_df.loc[combine_df['id']==31991]

index     id  label                                              tweet  \
31990  31990  31991    NaN   haroldfriday have a weekend filled with sunbe...   

                                             clean_tweet  \
31990  haroldfriday have weekend filled with sunbeams...   

                                             tweet_token  \
31990  [haroldfriday, have, weekend, filled, with, su...   

                                    tweet_token_filtered  \
31990  [haroldfriday, weekend, filled, sunbeams, ever...   

                                           tweet_stemmed  \
31990  [haroldfriday, weekend, fill, sunbeam, everyon...   

                                        tweet_lemmatized  
31990  [haroldfriday, weekend, fill, sunbeam, everyon...

ЗАДАНИЕ 1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

    ● Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
    ● Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
    ● Исключим стоп-слова с помощью stop_words='english'.
    ● Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [36]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd

In [34]:
tweet_stemmed_list=[]
for row in np.arange(len(combine_df)):
    tweet_stemmed_list.append(str(combine_df.at[row, 'tweet_stemmed']).replace('[','').replace(']','').replace('\'',''))

count_vectorizer_stemmed = CountVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_stemmed = count_vectorizer_stemmed.fit_transform(tweet_stemmed_list)

# Отобразим Bag-of-Words модель как DataFrame
feature_names_stemmed = count_vectorizer_stemmed.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns = feature_names_stemmed)


abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
49154    0        0       0        0    0       0      0       0   0      0   
49155    0        0       0        0    0       0      0       0   0      0   
49156    0        0       0        0    0       0      0       0   0      0   
49157    0        0       0        0    0       0      0       0   0      0   
49158    0        0       0        0    0       0      0       0   0      0   

       ...  yeah  year  yesterday  yo  yoga  york  young  youtub  yr  yummi  
0      ...     0     0          0   0     0     0      0       0   0      0  
1      ...     0     0          0   0     0     0      0       0   0      0  
2      ...     0     0          0   0     0     0      0       0   0      0  
3      ...     0     0          0   0     0     0      0       0   0      0  
4      ...     0     0          0   0     0     0      0       0   0      0  
...    ...   ...   ...        ...  ..   ...   ...    ...     ...  ..    ...  
49154  ...     0     0          0   0     0     0      0       0   0      0  
49155  ...     0     0          0   0     0     0      0       0   0      0  
49156  ...     0     0          0   0     0     0      0       0   0      0  
49157  ...     0     0          0   0     0     0      0       0   0      0  
49158  ...     0     0          0   0     0     0      0       0   0      0  

[49159 rows x 1000 columns]

In [35]:
tweet_lemmatized_list=[]
for row in np.arange(len(combine_df)):
    tweet_lemmatized_list.append(str(combine_df.at[row, 'tweet_lemmatized']).replace('[','').replace(']','').replace('\'',''))

count_vectorizer_lemmatized = CountVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_lemmatized = count_vectorizer_lemmatized.fit_transform(tweet_lemmatized_list)

# Отобразим Bag-of-Words модель как DataFrame
feature_names_lemmatized = count_vectorizer_lemmatized.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns = feature_names_lemmatized)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0         0           0       0        0    0       0      0         0      0   
1         0           0       0        0    0       0      0         0      0   
2         0           0       0        0    0       0      0         0      0   
3         0           0       0        0    0       0      0         0      0   
4         0           0       0        0    0       0      0         0      0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154     0           0       0        0    0       0      0         0      0   
49155     0           0       0        0    0       0      0         0      0   
49156     0           0       0        0    0       0      0         0      0   
49157     0           0       0        0    0       0      0         0      0   
49158     0           0       0        0    0       0      0         0      0   

       add  ...  yes  yesterday  yo  yoga  york  young  youth  youtube  yr  \
0        0  ...    0          0   0     0     0      0      0        0   0   
1        0  ...    0          0   0     0     0      0      0        0   0   
2        0  ...    0          0   0     0     0      0      0        0   0   
3        0  ...    0          0   0     0     0      0      0        0   0   
4        0  ...    0          0   0     0     0      0      0        0   0   
...    ...  ...  ...        ...  ..   ...   ...    ...    ...      ...  ..   
49154    0  ...    0          0   0     0     0      0      0        0   0   
49155    0  ...    0          0   0     0     0      0      0        0   0   
49156    0  ...    0          0   0     0     0      0      0        0   0   
49157    0  ...    0          0   0     0     0      0      0        0   0   
49158    0  ...    0          0   0     0     0      0      0        0   0   

       yummy  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
49154      0  
49155      0  
49156      0  
49157      0  
49158      0  

[49159 rows x 1000 columns]

ЗАДАНИЕ 2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

       ● Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
       ● Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
       ● Исключим стоп-слова с помощью stop_words='english'.
       ● Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


In [37]:
count_vectorizer_stemmed = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_stemmed = count_vectorizer_stemmed.fit_transform(tweet_stemmed_list)

# Отобразим Bag-of-Words модель как DataFrame
feature_names_stemmed = count_vectorizer_stemmed.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns = feature_names_stemmed)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  yeah  year  yesterday   yo  yoga  york  young  youtub   yr  yummi  
0      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
1      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
2      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
3      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
4      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
...    ...   ...   ...        ...  ...   ...   ...    ...     ...  ...    ...  
49154  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49155  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49156  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49157  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49158  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  

[49159 rows x 1000 columns]

In [38]:
count_vectorizer_lemmatized = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_lemmatized = count_vectorizer_lemmatized.fit_transform(tweet_lemmatized_list)

# Отобразим Bag-of-Words модель как DataFrame
feature_names_lemmatized = count_vectorizer_lemmatized.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns = feature_names_lemmatized)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
1       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
2       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
3       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
4       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49155   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49156   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49157   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49158   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   

       add  ...  yes  yesterday   yo  yoga  york  young  youth  youtube   yr  \
0      0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
1      0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
2      0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
3      0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
4      0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
...    ...  ...  ...        ...  ...   ...   ...    ...    ...      ...  ...   
49154  0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
49155  0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
49156  0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
49157  0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   
49158  0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0   

       yummy  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
49154    0.0  
49155    0.0  
49156    0.0  
49157    0.0  
49158    0.0  

[49159 rows x 1000 columns]

ЗАДАНИЕ 3. Натренируем gensim.models.Word2Vec модель на наших данных.

    ● Тренировать будем на токенизированных твитах combine_df['tweet_token']
    ● Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
    ● Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.


In [39]:
from gensim.models import Word2Vec, FastText

In [40]:
modelW2V = Word2Vec(sentences=combine_df['tweet_token'], size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)

In [56]:
modelW2V.train(combine_df['tweet_token'],total_examples=len(combine_df['tweet_token']),epochs=20)

(9075312, 11536060)

ЗАДАНИЕ 4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [58]:
result1=modelW2V.most_similar(positive=['dinner'])
print("{}: {:.4f}".format(*result1[0]))

result2=modelW2V.most_similar(positive=['trump'])
print("{}: {:.4f}".format(*result2[0]))

bihdaydinner: 0.5412
donald: 0.5694


ЗАДАНИЕ 5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".


In [64]:
vector=modelW2V['food']
vector

array([-3.19102071e-02, -2.57944137e-01,  8.81308019e-02, -4.23669755e-01,
        3.23475897e-02,  3.17507207e-01, -1.94420978e-01,  5.79137683e-01,
        2.78589517e-01, -1.31678283e-01,  1.47949353e-01, -6.33905977e-02,
        7.25393593e-01,  2.15091094e-01,  2.45795608e-01, -9.01857972e-01,
        2.50013024e-01,  1.06929874e+00, -2.29609281e-01,  3.37898463e-01,
        2.77195513e-01, -7.30865598e-02,  8.24091025e-04, -5.87754309e-01,
       -1.81916535e-01, -6.18198276e-01,  1.74955670e-02, -3.56160402e-01,
       -3.68711293e-01, -2.88392723e-01, -1.88813791e-01, -6.99500859e-01,
        2.47465014e-01, -4.04942125e-01,  5.85397542e-01, -2.30266824e-01,
        2.18938977e-01,  2.10888937e-01,  6.11920178e-01, -1.78562596e-01,
        4.55489039e-01,  4.42466140e-01,  2.04333067e-01,  3.05638790e-01,
       -1.50496423e-01,  7.95378625e-01, -3.28240767e-02,  3.72092068e-01,
        6.47284985e-01, -6.48265779e-01, -1.32418960e-01, -4.50770736e-01,
        2.78406024e-01, -

ЗАДАНИЕ 6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).


In [79]:
size=200
combine_df['tweet_vecs']=''
for row in np.arange(len(combine_df)):
    vector_w2v = np.zeros(size).reshape((1, size))
    n_w2v = 0
    for word in combine_df.at[row, 'tweet_token']:
        if word in modelW2V:
            vector_w2v += modelW2V[word].reshape((1, size))
            n_w2v += 1
    if n_w2v > 0:
        vector_w2v = vector_w2v / n_w2v
        combine_df.at[row,'tweet_vecs']=vector_w2v
        

In [83]:
combine_df.at[0,'tweet_vecs'].shape

(1, 200)